scripts/mech_events.py

```diff
@@ -0,0 +1,266 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+# ------------------------------------------------------------------------------
+#
+#   Copyright 2022-2023 Valory AG
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+
+"""Utilities to retrieve on-chain Mech events."""
+
+import json
+import time
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Any, Dict
+
+import requests
+from eth_utils import to_checksum_address
+from tqdm import tqdm
+from web3 import HTTPProvider, Web3
+from web3.datastructures import AttributeDict
+from web3.types import BlockParams
+
+
+SCRIPT_PATH = Path(__file__).resolve().parent
+AGENT_MECH_JSON_PATH = Path(SCRIPT_PATH, "..", "contracts", "AgentMech.json")
+MECH_EVENTS_JSON_PATH = Path(SCRIPT_PATH, "..", ".trader_runner", "mech_events.json")
+HTTP = "http://"
+HTTPS = HTTP[:4] + "s" + HTTP[4:]
+CID_PREFIX = "f01701220"
+IPFS_ADDRESS = f"{HTTPS}gateway.autonolas.tech/ipfs/"
+LATEST_BLOCK_NAME: BlockParams = "latest"
+BLOCK_DATA_NUMBER = "number"
+BLOCKS_CHUNK_SIZE = 5000
+EXCLUDED_BLOCKS_THRESHOLD = 2 * BLOCKS_CHUNK_SIZE
+NUM_EXCLUDED_BLOCKS = 10
+
+# Pair of Mech contract address and the earliest block to search events
+# (either where the block where the contract was created, or where the
+# first event was recorded on-chain)
+MECH_CONTRACT_ADDRESSES = [
+    # Old Mech contract
+    (
+        to_checksum_address("0xff82123dfb52ab75c417195c5fdb87630145ae81"),
+        28911547,
+    ),
+    # New Mech contract
+    (
+        to_checksum_address("0x77af31de935740567cf4ff1986d04b2c964a786a"),
+        30663133,
+    ),
```

Comment:
The Deliver events in these two contracts have different signatures. You will need both ABIs to get Delivers from these two contracts. This is how its done in the mech:
https://github.com/valory-xyz/mech/blob/main/packages/valory/contracts/agent_mech/contract.py#L201

Reply:
We are not retrieving Deliver events in this PR.

Reply:
Okay the `MechDeliver` dataclass gave me the impression its used somewhere.

Reply:
Removed the class MechDeliver. If needed in future PRs can be re-added, and the issue you raised addressed.